modules/supervisord.nix

```diff
@@ -0,0 +1,177 @@
+  });
+
+  renderAtom = val:
+    if builtins.isBool val then if val then "true" else "false"
+    else toString val;
+
+  renderProgram = program: let
+    section = {
+      inherit (program) command;
+      autorestart = program.autoRestart;
+      environment = let
+        # FIXME: Make more robust
+        escape = builtins.replaceStrings [ "%" ] [ "%%" ];
+        envs = lib.mapAttrsToList (k: v: "${k}=\"${escape v}\"") program.environment;
+      in builtins.concatStringsSep "," envs;
+    } // program.extraConfig;
+  in lib.mapAttrs (_: v: renderAtom v) section;
+
+  numPrograms = builtins.length (builtins.attrNames cfg.programs);
+  enabledPrograms = lib.filterAttrs (_: program: program.enable) cfg.programs;
+
+  structuredConfig = {
+    supervisord = {
+      logfile = cfg.logPath;
+      pidfile = cfg.pidPath;
+    };
+    supervisorctl = {
+      serverurl = "unix://${cfg.socketPath}";
+    };
+    unix_http_server = {
+      file = cfg.socketPath;
+    };
+    "rpcinterface:supervisor" = {
+      "supervisor.rpcinterface_factory" = "supervisor.rpcinterface:make_main_rpcinterface";
+    };
+  } // (lib.mapAttrs' (k: v: {
+    name = "program:${k}";
+    value = renderProgram v;
+  }) enabledPrograms);
+
+  configFile = format.generate "supervisord.conf" structuredConfig;
+
+  # Only expose the "supervisorctl" executable
+  supervisorctl = pkgs.runCommand "supervisorctl" {} ''
+    mkdir -p $out/bin
+    ln -s ${cfg.package}/bin/supervisorctl $out/bin/supervisorctl
+  '';
+in {
+  options = {
+    supervisord = {
+      enable = lib.mkOption {
+        description = ''
+          Whether to enable the supervisord process control system.
+
+          This allows you to define long-running services in Nix-on-Droid.
+        '';
+        type = types.bool;
+        default = numPrograms != 0;
+      };
+      package = lib.mkOption {
+        description = ''
+          The supervisord package to use.
+        '';
+        type = types.package;
+        default = pkgs.python3Packages.supervisor;
+        defaultText = lib.literalExpression "pkgs.python3Packages.supervisor";
+      };
+      socketPath = lib.mkOption {
+        description = ''
+          Path to the UNIX domain socket on which supervisord will listen on.
+        '';
+        type = types.path;
+        default = "/tmp/supervisor.sock";
+      };
+      pidPath = lib.mkOption {
+        description = ''
+          Path to the file in which supervisord saves its PID.
+        '';
+        type = types.path;
+        default = "/tmp/supervisor.pid";
+      };
+      logPath = lib.mkOption {
+        description = ''
+          Path to the log file.
+        '';
+        type = types.path;
+        default = "/tmp/supervisor.log";
+      };
+      programs = lib.mkOption {
+        description = ''
+          Definition of supervisord programs.
+        '';
+        type = types.attrsOf programType;
+        default = {};
+      };
+      configFile = lib.mkOption {
```

Comment:
Why do you define an option for this? Do you have a use case for this option in mind? Otherwise, I would prefer to just use the `configFile` variable without this indirection.

Reply:
I initially wanted to use the hardcoded config path to start supervisord itself in `login-inner`, but reloading is broken this way (what did I expect :stuck_out_tongue:). So this isn't really useful anymore and let's remove it.